completion_aggregator/views.py

```diff
@@ -0,0 +1,54 @@
+"""
+completion_aggregator App progress bar view
+"""
+from __future__ import absolute_import, unicode_literals
+
+from django.conf import settings
+from django.contrib.auth.mixins import LoginRequiredMixin
+from django.shortcuts import render
+from django.views.decorators.clickjacking import xframe_options_exempt
+from django.views.generic import TemplateView
+from xblockutils.resources import ResourceLoader
+
+from .api.v1.views import CompletionDetailView
+
+loader = ResourceLoader(__name__)
+
+
+class CompletionProgressBarView(LoginRequiredMixin, TemplateView):
```

Comment:
How about the following refactor of this logic?

```python
class CompletionProgressBarView(LoginRequiredMixin, TemplateView):
    """
    View to display the progress bar of a student in a course
    """
    def __get_user_completion(username, chapter_id, results):
        """
        Return the user completion percentage, using the completion response.

        In case the user completion cannot be returned as a result of missing user completion, we return None,
        indicating its absence.
        """
        user_completion = next(filter(results, lambda r: r.get('username') == username), None)

        # No completion returned, hence we cannot get the percentage either
        # Indicate no user completion by returning None
        if not user_completion:
            return None

        chapters = user_completion['chapter']
        chapter = next(filter(chapters, lambda c: c['block_key'].split('@')[-1] == chapter_id), None)

        completion_kind = chapter if chapter else user_completion
        return float(completion_kind['completion']['percent']) * 100

    @xframe_options_exempt
    def get(self, request, course_key, chapter_id=None):
        """
        Fetch progress and render the template.
        """
        completion_percentage = 0
        completion_resp = CompletionDetailView.as_view()(request, course_key).data

        if chapter_id is not None:
            new_req = request.GET.copy()
            new_req['requested_fields'] = "chapter"
            request.GET = new_req

        if completion_resp:
            results = completion_resp.get('results')
            user_completion_percentage = self.__get_user_completion(request.user.username, chapter_id, results)

            if user_completion_percentage:
                completion_percentage = user_completion_percentage

        template = 'chapter_completion_progress_bar.html' if chapter_id is not None else 'completion_progress_bar.html'
        return render(request, template, {
            'completion_percentage': completion_percentage,
        })
```

Reply:
Also, could you please add tests for the view?